## Test Cricket Bowlers Analysis
Author: Mehreen Tabassum<br>
EEE, RUET
### Dataset sourse- https://stats.espncricinfo.com/ci/content/records/93276.html

#### Importing the libraries

In [1]:
import pandas as pd
import numpy as np

#### Reading 'wickets' sheet from 'test_cricket.xtsx' 

In [2]:
df = pd.read_excel("test_cricket.xlsx", sheet_name='wickets')

#### First 10 rows of the dataset

In [3]:
display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   

                   BBI                  BBM    Ave  Econ    SR   5  10  
0  1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1  1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2  1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3  1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4  2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5  2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6  1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7  1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8  1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   2  
9                9/127               14/184  28.07  2.80  60.0  34   9

#### Number of rows and column 

In [4]:
print("Number of rows -",df.shape[0])
print("Number of columns -",df.shape[1])

Number of rows - 79
Number of columns - 14


#### Present Columns and their meaning

In [5]:
print(df.columns)

Index(['Player',   'Span',    'Mat',   'Inns',  'Balls',   'Runs',   'Wkts',
          'BBI',    'BBM',    'Ave',   'Econ',     'SR',        5,       10],
      dtype='object')


##### Column names 
1. **Player** - Name of the players who played test cricket
2. **Span** - Career span of each player
3. **Mat** - Number of matches played
4. **Inns** - Total innings played by the player
5. **Balls** - Total number of balls bowled by the player
6. **Runs** - The number of runs conceded by the player
7. **Wkts** - Total number of wickets taken by the player
8. **BBI** - Best bowling in an innings of the player	
9. **BBM** - Best bowling in a match of the player
10. **Ave** - The average number of runs conceded per wicket by the player. (Runs/W)
11. **Econ** - The average number of runs conceded per over. (Runs/Overs bowled)
12. **SR** - The average number of balls bowled per wicket taken. (Balls/W)
13. **5** - The number of innings in which the bowler took at least five wickets.
14. **10** - The number of matches in which the bowler took at least ten wickets.

#### Statistical measures from the dataset

In [6]:
display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

##### Statistic measures
- Count- total numbers of observation
- mean- average of the values
- std- standard deviation
- min- minimum value in the column
- 25%- 25% of all value is under the certain column value
- 50%- 50% of all value is under the certain column value
- 75%- 75% of all value is under the certain column value
- max- maximum value in the column

##### Observations
- On an average a played has played 80 matches
- Minimum 37 matches were played by a player
- 25% of players have played 60 matches or less
- 50% of players have played 71 matches or less
- 75% of players have played 93 matches or less
- Maximum 166 matches were played by a player
- Maximum 800 wickets were taken by a player
- Minimum 67 innings were played by a player

Many other primary insights can be taken from the the statistical measures

#### Renaming the columns

In [7]:
df = df.rename(columns={'Mat':'Match',
                        'Inns':'Innings',
                        'Wkts':'Wickets',
                        'BBI':'Best Bowling(Innings)',
                        'BBM':'Best Bowling(Match)',
                        'Ave':'Bowling avg',
                        'Econ':'Economy rate',
                        'SR':'Strike rate',
                         5:'5 wickets',
                         10:'10 wickets'})
display(df.head(3))

Player       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010    133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007    145      273  40705  17995      708   
2         A Kumble (INDIA)  1990-2008    132      236  40850  18355      619   

  Best Bowling(Innings) Best Bowling(Match)  Bowling avg  Economy rate  \
0   1951-09-01 00:00:00              16/220        22.72          2.47   
1   1971-08-01 00:00:00              12/128        25.41          2.65   
2   1974-10-01 00:00:00              14/149        29.65          2.69   

   Strike rate  5 wickets  10 wickets  
0         55.0         67          22  
1         57.4         37          10  
2         65.9         35           8

#### Datatypes of the variables in the coulmn

In [8]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Player                 79 non-null     object 
 1   Span                   79 non-null     object 
 2   Match                  79 non-null     int64  
 3   Innings                79 non-null     int64  
 4   Balls                  79 non-null     int64  
 5   Runs                   79 non-null     int64  
 6   Wickets                79 non-null     int64  
 7   Best Bowling(Innings)  79 non-null     object 
 8   Best Bowling(Match)    79 non-null     object 
 9   Bowling avg            79 non-null     float64
 10  Economy rate           79 non-null     float64
 11  Strike rate            79 non-null     float64
 12  5 wickets              79 non-null     int64  
 13  10 wickets             79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB
Non

##### Observations:

- Number of entries(rows) are 79 and and 14 columns.<br>
- There are-
> **3 Float** type variables- Bowling average, Economy rate, Strike rate<br>
> **4 String** type variables- Best Bowling (Innings), Best Bowling (Match), Player, Span<br>
> **7 Integer** type variables- Match, Innings, Balls, Runs, Wickets, 5 wickets (innings), 10 wickets (match)

- From the non-null columns, we can see that there are **no missing value**
---

#### Droping Best Bowling(Innings) and Best Bowling(Match) column

Few data in the column BBI and BBM have got formatted as date instead of number in the format of 'xx/xxx'.<br>
Removing BBI, BBM column 

In [9]:
df.drop(['Best Bowling(Innings)','Best Bowling(Match)'],axis=1 , inplace=True)
display(df.head(5))

Player       Span  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010    133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007    145      273  40705  17995      708   
2         A Kumble (INDIA)  1990-2008    132      236  40850  18355      619   
3        JM Anderson (ENG)  2003-2021    162      301  34791  16457      617   
4         GD McGrath (AUS)  1993-2007    124      243  29248  12186      563   

   Bowling avg  Economy rate  Strike rate  5 wickets  10 wickets  
0        22.72          2.47         55.0         67          22  
1        25.41          2.65         57.4         37          10  
2        29.65          2.69         65.9         35           8  
3        26.67          2.83         56.3         30           3  
4        21.64          2.49         51.9         29           3

#### Main dataframe after Splitting Player name and Country 

In [10]:
#removing the closing bracket
df['Player'] = df['Player'].str.replace(")"," ", regex=True)  

#splitting
Player_country=df['Player'].str.split("(", expand=True)  

#### Splitted column

In [11]:
display(Player_country.head())

0        1
0  M Muralitharan   ICC/SL 
1        SK Warne      AUS 
2        A Kumble    INDIA 
3     JM Anderson      ENG 
4      GD McGrath      AUS

#### Dropping player column

In [12]:
df.drop('Player', axis=1 ,inplace=True)  
display(df.head())

Span  Match  Innings  Balls   Runs  Wickets  Bowling avg  \
0  1992-2010    133      230  44039  18180      800        22.72   
1  1992-2007    145      273  40705  17995      708        25.41   
2  1990-2008    132      236  40850  18355      619        29.65   
3  2003-2021    162      301  34791  16457      617        26.67   
4  1993-2007    124      243  29248  12186      563        21.64   

   Economy rate  Strike rate  5 wickets  10 wickets  
0          2.47         55.0         67          22  
1          2.65         57.4         37          10  
2          2.69         65.9         35           8  
3          2.83         56.3         30           3  
4          2.49         51.9         29           3

#### Reorganizing the dataframe after splitting country

In [13]:
df = pd.concat([df, Player_country], axis=1)  #concatenating the splitted column with main dataframe

df = df.rename(columns={0: 'Player',1: 'Country'})  #renaming the columns

column_sequence = ['Player', 'Span', 'Country', 'Match', 'Innings', 'Balls', 'Runs','Wickets',
                   'Bowling avg','Economy rate','Strike rate', '5 wickets', '10 wickets']   #new column sequence as a list

df = df[column_sequence]  #implementing the new column sequence

display(df.head())

Player       Span  Country  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan   1992-2010  ICC/SL     133      230  44039  18180      800   
1        SK Warne   1992-2007     AUS     145      273  40705  17995      708   
2        A Kumble   1990-2008   INDIA     132      236  40850  18355      619   
3     JM Anderson   2003-2021     ENG     162      301  34791  16457      617   
4      GD McGrath   1993-2007     AUS     124      243  29248  12186      563   

   Bowling avg  Economy rate  Strike rate  5 wickets  10 wickets  
0        22.72          2.47         55.0         67          22  
1        25.41          2.65         57.4         37          10  
2        29.65          2.69         65.9         35           8  
3        26.67          2.83         56.3         30           3  
4        21.64          2.49         51.9         29           3

#### ICC Players

##### Function for checking players played for ICC

In [14]:
def ICC(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

***A new column created while applying the function***

In [15]:
# using ICC function to check
df['Played_for_ICC'] = df['Country'].apply(ICC)

# removing "ICC/" from string of country column
df['Country'] = df['Country'].str.replace("ICC/","", regex=True)
display(df.head(5))

Player       Span Country  Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan   1992-2010     SL     133      230  44039  18180      800   
1        SK Warne   1992-2007    AUS     145      273  40705  17995      708   
2        A Kumble   1990-2008  INDIA     132      236  40850  18355      619   
3     JM Anderson   2003-2021    ENG     162      301  34791  16457      617   
4      GD McGrath   1993-2007    AUS     124      243  29248  12186      563   

   Bowling avg  Economy rate  Strike rate  5 wickets  10 wickets  \
0        22.72          2.47         55.0         67          22   
1        25.41          2.65         57.4         37          10   
2        29.65          2.69         65.9         35           8   
3        26.67          2.83         56.3         30           3   
4        21.64          2.49         51.9         29           3   

  Played_for_ICC  
0            Yes  
1             No  
2             No  
3             No  
4             No

***Counting ICC players***

In [16]:
print(df['Played_for_ICC'].value_counts())

No     74
Yes     5
Name: Played_for_ICC, dtype: int64


##### Observation
- Only **5** players played for **ICC**

#### Different countries

In [17]:
# no. of countries
diff_countries=len(df['Country'].unique()) 

# print(diff_countries, "different countries are present in the dataset")

> **11 different countries** are present in the dataset

#### Aussie bowlers: Bangladeshi bowlers

In [18]:
print(df['Country'].value_counts())

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
PAK          7
NZ           7
SL           3
ENG/ICC      2
BDESH        1
ZIM          1
Name: Country, dtype: int64


##### Observation
- In the dataset there are **18 Australian** bowlers but only **1 Bangladeshi** bowler.

#### Reorganizing the dataframe after splitting start and end year

In [19]:
# splitting 'Span' column
df_span = df['Span'].str.split("-", expand=True)

# concatenating the new dataframe with the main dataframe
df = pd.concat([df, df_span], axis=1)

# renaming new splitted column
df = df.rename(columns={0: "Start_year",
                        1: "End_year"})

# removing 'Span' column
df = df.drop("Span", axis=1)

#new column sequence as a list
column_sequence = ['Player', 'Country','Start_year', 'End_year', 'Match', 'Innings', 'Balls', 'Runs','Wickets',
                   'Bowling avg','Economy rate','Strike rate', '5 wickets', '10 wickets']   

#implementing the new column sequence
df = df[column_sequence]  

display(df.head())

Player Country Start_year End_year  Match  Innings  Balls   Runs  \
0  M Muralitharan      SL        1992     2010    133      230  44039  18180   
1        SK Warne     AUS        1992     2007    145      273  40705  17995   
2        A Kumble   INDIA        1990     2008    132      236  40850  18355   
3     JM Anderson     ENG        2003     2021    162      301  34791  16457   
4      GD McGrath     AUS        1993     2007    124      243  29248  12186   

   Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  10 wickets  
0      800        22.72          2.47         55.0         67          22  
1      708        25.41          2.65         57.4         37          10  
2      619        29.65          2.69         65.9         35           8  
3      617        26.67          2.83         56.3         30           3  
4      563        21.64          2.49         51.9         29           3

#### Total years of career

In [20]:
# changing data type of the starting and ending year to integer 
# df['Start_year'] = df['Start_year'].astype('int') - another method

df['End_year']=pd.to_numeric(df['End_year'])
df['Start_year']=pd.to_numeric(df['Start_year'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Player        79 non-null     object 
 1   Country       79 non-null     object 
 2   Start_year    79 non-null     int64  
 3   End_year      79 non-null     int64  
 4   Match         79 non-null     int64  
 5   Innings       79 non-null     int64  
 6   Balls         79 non-null     int64  
 7   Runs          79 non-null     int64  
 8   Wickets       79 non-null     int64  
 9   Bowling avg   79 non-null     float64
 10  Economy rate  79 non-null     float64
 11  Strike rate   79 non-null     float64
 12  5 wickets     79 non-null     int64  
 13  10 wickets    79 non-null     int64  
dtypes: float64(3), int64(9), object(2)
memory usage: 8.8+ KB


> The starting and ending years columns are **changed from object to integer variable**; <br>
**Suitable for mathematical operations**; To find the total years of player's career.

In [21]:
#total years of career of the players
df['Career_span']=df['End_year']-df['Start_year']

#rearranging columns
df = df[['Player', 'Country','Start_year', 'End_year','Career_span','Match', 'Innings', 'Balls', 'Runs',
         'Wickets', 'Bowling avg','Economy rate','Strike rate', '5 wickets', '10 wickets']] 
display(df.head())

Player Country  Start_year  End_year  Career_span  Match  Innings  \
0  M Muralitharan      SL         1992      2010           18    133      230   
1        SK Warne     AUS         1992      2007           15    145      273   
2        A Kumble   INDIA         1990      2008           18    132      236   
3     JM Anderson     ENG         2003      2021           18    162      301   
4      GD McGrath     AUS         1993      2007           14    124      243   

   Balls   Runs  Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  \
0  44039  18180      800        22.72          2.47         55.0         67   
1  40705  17995      708        25.41          2.65         57.4         37   
2  40850  18355      619        29.65          2.69         65.9         35   
3  34791  16457      617        26.67          2.83         56.3         30   
4  29248  12186      563        21.64          2.49         51.9         29   

   10 wickets  
0          22  
1          10  
2           8  
3           3  
4           3

#### Players with longest and shortest career span

In [22]:
print(df['Career_span'].value_counts())

10    11
13    10
9      9
14     8
12     7
11     6
15     6
17     5
18     5
7      3
8      2
16     2
19     1
20     1
5      1
6      1
21     1
Name: Career_span, dtype: int64


> The **longest (21 yrs)** and the **shortest (5 yrs) career** is of **1 player each**

In [23]:
display(df.sort_values(by='Career_span', ascending = False).head(1))

Player Country  Start_year  End_year  Career_span  Match  Innings  \
21  Imran Khan     PAK         1971      1992           21     88      142   

    Balls  Runs  Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  \
21  19458  8258      362        22.81          2.54         53.7         23   

    10 wickets  
21           6

> **Imran Khan** has the **longest career** of **21 years**

In [24]:
display(df.sort_values(by='Career_span', ascending = True).head(1))

Player Country  Start_year  End_year  Career_span  Match  Innings  \
44  GP Swann     ENG         2008      2013            5     60      109   

    Balls  Runs  Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  \
44  15349  7642      255        29.96          2.98         60.1         17   

    10 wickets  
44           3

> **GP Swann** has the **shortest career** of **5 years**

#### Lowest economy rate

In [25]:
display(df.sort_values(by='Economy rate', ascending = True).head(3))

Player Country  Start_year  End_year  Career_span  Match  Innings  \
32      LR Gibbs      WI         1958      1976           18     79      148   
47      R Benaud     AUS         1952      1964           12     63      116   
35  DL Underwood     ENG         1966      1982           16     86      151   

    Balls  Runs  Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  \
32  27115  8989      309        29.09          1.98         87.7         18   
47  19108  6704      248        27.03          2.10         77.0         16   
35  21862  7674      297        25.83          2.10         73.6         17   

    10 wickets  
32           2  
47           1  
35           6

> **GP Swann** has the **lowest economy rate** of **1.98**

#### Lowest strike rate

In [26]:
display(df.sort_values(by='Strike rate', ascending = True).head(3))

Player Country  Start_year  End_year  Career_span  Match  Innings  \
71      K Rabada      SA         2015      2021            6     47       86   
7       DW Steyn      SA         2004      2019           15     93      171   
20  Waqar Younis     PAK         1989      2003           14     87      154   

    Balls   Runs  Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  \
71   8785   4846      213        22.75          3.30         41.2         10   
7   18608  10077      439        22.95          3.24         42.3         26   
20  16224   8788      373        23.56          3.25         43.4         22   

    10 wickets  
71           4  
7            5  
20           5

> **K Rabada** has the **lowest strike rate** of **41.2**

#### Lowest bowling average 

In [27]:
display(df.sort_values(by='Bowling avg', ascending = True).head(3))

Player Country  Start_year  End_year  Career_span  Match  Innings  \
19  MD Marshall      WI         1978      1991           13     81      151   
41     J Garner      WI         1977      1987           10     58      111   
15  CEL Ambrose      WI         1988      2000           12     98      179   

    Balls  Runs  Wickets  Bowling avg  Economy rate  Strike rate  5 wickets  \
19  17584  7876      376        20.94          2.68         46.7         22   
41  13169  5433      259        20.97          2.47         50.8          7   
15  22103  8501      405        20.99          2.30         54.5         22   

    10 wickets  
19           4  
41           0  
15           3

> **MD Marshall** has the **lowest bowling average** of **20.94**
---